### PR TITLE
James 2746 - slow JMAP getMailboxes

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Quotas.java [hunk 1]
```diff
@@ -66,6 +67,20 @@ private QuotaId(QuotaRoot quotaRoot) {
         public String getName() {
             return quotaRoot.getValue();
         }
+
+        @Override
+        public boolean equals(Object o) {
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Quotas.java]: can you add EqualsVerifier test for this?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 2]
```diff
@@ -121,31 +124,45 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
...
+    private MailboxFactory.MailboxBuilder getBuilder(
+        MailboxSession mailboxSession,
+        MailboxId mailboxId,
+        Optional<List<MailboxMetaData>> preloadedMailboxMetadata,
+        Optional<Quotas> defaultQuotas) {
+        final MailboxFactory.MailboxBuilder builder = mailboxFactory.builder()
+            .id(mailboxId)
+            .session(mailboxSession);
+        final MailboxFactory.MailboxBuilder builderWithQuotas = defaultQuotas.map(builder::usingPreloadedUserDefaultQuotas).orElse(builder);
+        return preloadedMailboxMetadata.map(builderWithQuotas::usingPreloadedMailboxesMetadata).orElse(builderWithQuotas);
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: how about replacing .orElse() by lazily .orElseGet()

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java [hunk 3]
```diff
@@ -138,15 +147,24 @@ private Mailbox fromMessageManager(MessageManager messageManager, Optional<List<
...
-            QuotaId.fromQuotaRoot(quotaRoot),
+            quotaId,
             Quotas.Quota.from(
                 quotaToValue(quotaManager.getStorageQuota(quotaRoot)),
                 quotaToValue(quotaManager.getMessageQuota(quotaRoot))));
     }
 
+    public Quotas getUserDefaultQuotas(MailboxSession mailboxSession) throws MailboxException {
+        MailboxPath inboxPath = MailboxPath.inbox(mailboxSession);
+        return (getQuotas(inboxPath, Optional.empty()));
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: s/(getQuotas(inboxPath, Optional.empty()))/getQuotas(inboxPath, Optional.empty()) ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 4]
```diff
@@ -121,31 +124,45 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
...
 
+    private MailboxFactory.MailboxBuilder getBuilder(
+        MailboxSession mailboxSession,
+        MailboxId mailboxId,
+        Optional<List<MailboxMetaData>> preloadedMailboxMetadata,
+        Optional<Quotas> defaultQuotas) {
+        final MailboxFactory.MailboxBuilder builder = mailboxFactory.builder()
+            .id(mailboxId)
+            .session(mailboxSession);
+        final MailboxFactory.MailboxBuilder builderWithQuotas = defaultQuotas.map(builder::usingPreloadedUserDefaultQuotas).orElse(builder);
```
[the remark this answers is not in the document]
Comment (remk) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: As the builder is already instantiated, i don't see the need to use the lazy method

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 5]
```diff
@@ -139,6 +139,7 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
     private Stream<Mailbox> retrieveSpecificMailboxes(MailboxSession mailboxSession, ImmutableList<MailboxId> mailboxIds) {
         return mailboxIds
             .stream()
+            .parallel()
```
[the remark this answers is not in the document]
Comment (remk) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: good idea

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java (hunk 2 above)
Comment (remk) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: As the builder is already instantiated, i don't see the need to use the lazy method

Comment (blackheaven) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: add CR please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (blackheaven) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: We should use Reactor, Java's `parallel` is buggy.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 6]
```diff
@@ -121,31 +124,47 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
...
+        Optional<Quotas> defaultQuotas) {
+        MailboxFactory.MailboxBuilder builder = mailboxFactory.builder()
+            .id(mailboxId)
+            .session(mailboxSession);
+        MailboxFactory.MailboxBuilder builderWithQuotas = defaultQuotas.map(builder::usingPreloadedUserDefaultQuotas).orElse(builder);
+        return preloadedMailboxMetadata.map(builderWithQuotas::usingPreloadedMailboxesMetadata).orElse(builderWithQuotas);
+    }
+
     private Stream<Mailbox> retrieveSpecificMailboxes(MailboxSession mailboxSession, ImmutableList<MailboxId> mailboxIds) {
         return mailboxIds
```
Comment (blackheaven) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: Start the `Flux` from the beginning

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 7]
```diff
@@ -121,31 +124,47 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
...
-            MailboxQuery.builder()
-                .matchesAllMailboxNames()
-                .build(),
-            mailboxSession);
+        List<MailboxMetaData> userMailboxes = getAllMailboxesMetaData(mailboxSession);
+        Quotas inboxQuotas =  mailboxFactory.getUserDefaultQuotas(mailboxSession);
+
         return userMailboxes
             .stream()
+            .parallel()
```
Comment (blackheaven) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: Use `Flux` here also

**Comment by remk** [comment]
Before : 

[getMailboxes_master.pdf](https://github.com/linagora/james-project/files/3113231/getMailboxes_master.pdf)
After : 

[getMailboxes_2746.pdf](https://github.com/linagora/james-project/files/3113230/getMailboxes_2746.pdf)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 8]
```diff
@@ -41,19 +42,25 @@
...
+import reactor.core.publisher.ParallelFlux;
+import reactor.core.scheduler.Schedulers;
 
 public class GetMailboxesMethod implements Method {
 
     private static final Method.Request.Name METHOD_NAME = Method.Request.name("getMailboxes");
     private static final Method.Response.Name RESPONSE_NAME = Method.Response.name("mailboxes");
+    private static final Optional<List<MailboxMetaData>> NO_PRELOADED_METADATA = Optional.empty();
+    private static final Optional<Quotas> NO_PRELOADED_QUOTAS = Optional.empty();
+    public static final int PARALLELISM = 8;
```
Comment (blackheaven) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: `MAILBOX_FETCHING_PARALLELISM`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java [hunk 9]
```diff
@@ -138,19 +147,28 @@ private Mailbox fromMessageManager(MessageManager messageManager, Optional<List<
             .build();
     }
 
-    private Quotas getQuotas(MailboxPath mailboxPath) throws MailboxException {
+    private Quotas getQuotas(MailboxPath mailboxPath, Optional<Quotas> preloadedUserDefaultQuotas) throws MailboxException {
         QuotaRoot quotaRoot = quotaRootResolver.getQuotaRoot(mailboxPath);
+        QuotaId quotaId = QuotaId.fromQuotaRoot(quotaRoot);
+        if (preloadedUserDefaultQuotas.isPresent() && preloadedUserDefaultQuotas.get().getQuotas().containsKey(quotaId)) {
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: seems like optional is making the if check becomes harder to read,  I propose:
```java
        boolean containsQuotaId = preloadedUserDefaultQuotas
            .map(Quotas::getQuotas)
            .map(quotaIdQuotaMap -> quotaIdQuotaMap.containsKey(quotaId))
            .orElse(false);

        if (containsQuotaId) {
            return preloadedUserDefaultQuotas.get();
        }
```

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 10]
```diff
@@ -90,62 +97,82 @@
...
     }
 
     private Set<MailboxProperty> ensureContainsId(Set<MailboxProperty> input) {
         return Sets.union(input, ImmutableSet.of(MailboxProperty.ID)).immutableCopy();
     }
 
-    private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesRequest, MailboxSession mailboxSession) {
-        GetMailboxesResponse.Builder builder = GetMailboxesResponse.builder();
+    private Mono<GetMailboxesResponse> getMailboxesResponse(GetMailboxesRequest mailboxesRequest, MailboxSession mailboxSession) {
+        final GetMailboxesResponse.Builder builder = GetMailboxesResponse.builder();
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: you don't need final keyword, right?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 11]
```diff
@@ -90,62 +97,82 @@
...
         try {
             Optional<ImmutableList<MailboxId>> mailboxIds = mailboxesRequest.getIds();
-            retrieveMailboxes(mailboxIds, mailboxSession)
-                .sorted(Comparator.comparing(Mailbox::getSortOrder))
-                .forEach(builder::add);
-            return builder.build();
+            return retrieveMailboxes(mailboxIds, mailboxSession)
```
Comment (trantienduchn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]:
```java
            return retrieveMailboxes(mailboxesRequest.getIds(), mailboxSession)
                .collectSortedList(Comparator.comparing(Mailbox::getSortOrder))
                .map(mailboxes -> GetMailboxesResponse.builder().addAll(mailboxes).build());
```
WDYT?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 12]
```diff
@@ -90,62 +97,82 @@
...
+    private MailboxFactory.MailboxBuilder getBuilder(
+        MailboxSession mailboxSession,
+        MailboxId mailboxId,
+        Optional<List<MailboxMetaData>> preloadedMailboxMetadata,
+        Optional<Quotas> defaultQuotas) {
+        MailboxFactory.MailboxBuilder builder = mailboxFactory.builder()
+            .id(mailboxId)
+            .session(mailboxSession);
+        MailboxFactory.MailboxBuilder builderWithQuotas = defaultQuotas
+            .map(builder::usingPreloadedUserDefaultQuotas)
```
[the remark this answers is not in the document]
Comment (remk) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: I am not sure. Even if it's nicer to use in our case. I don't think it make sense to accept Optional in a builder.

**Comment by trantienduchn** [comment]
```
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] [ERROR]   CassandraMessageIdToImapUidDAOTest.updateShouldUpdateDeletedFlag:253 
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] Expecting:
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=ebf2f491-6730-11e9-856f-43751af5c660}, messageId=CassandraMessageId{uuid=ebf2f490-6730-11e9-856f-43751af5c660}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] to contain only:
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=ebf2f491-6730-11e9-856f-43751af5c660}, messageId=CassandraMessageId{uuid=ebf2f490-6730-11e9-856f-43751af5c660}, uid=MessageUid{uid=1}}, flags=\Deleted, modSeq=2}]>
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] elements not found:
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=ebf2f491-6730-11e9-856f-43751af5c660}, messageId=CassandraMessageId{uuid=ebf2f490-6730-11e9-856f-43751af5c660}, uid=MessageUid{uid=1}}, flags=\Deleted, modSeq=2}]>
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] and elements not expected:
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=ebf2f491-6730-11e9-856f-43751af5c660}, messageId=CassandraMessageId{uuid=ebf2f490-6730-11e9-856f-43751af5c660}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[0eb2001fa667f0e32cbb060cd567f1e95ff35a07] 
```

**Comment by remk** [comment]
Result for the Quotas optimization without parallelism.

Capture d’écran de 2019-04-25 11-13-10

It's sad, but the parallel fetching execution of the mailboxes gives us nothing except an higher errors rate....

**Comment by trantienduchn** [comment]
> Result for the Quotas optimization without parallelism.
> 
> Capture d’écran de 2019-04-25 11-13-10
> 
> It's sad, but the parallel fetching execution of the mailboxes gives us nothing except an higher errors rate....

Sorry, what is this implementation? The implementation includes [this](https://github.com/linagora/james-project/pull/2324#discussion_r278389188) ?

**Comment by trantienduchn** [comment]
> Before :
> 
> [getMailboxes_master.pdf](https://github.com/linagora/james-project/files/3113231/getMailboxes_master.pdf)
> After :
> 
> [getMailboxes_2746.pdf](https://github.com/linagora/james-project/files/3113230/getMailboxes_2746.pdf)

Now I just realize that the after PDF has error while the master PDF doesn't have any error.

**Comment by remk** [comment]
> > Result for the Quotas optimization without parallelism.
> > Capture d’écran de 2019-04-25 11-13-10
> > It's sad, but the parallel fetching execution of the mailboxes gives us nothing except an higher errors rate....
> 
> Sorry, what is this implementation? The implementation includes [this](https://github.com/linagora/james-project/pull/2324#discussion_r278389188) ?
No this implementation only use the commits dedicated to the quotas fetching optimization i.e
commit : James 2746 - slow JMAP getMailboxes and it's fixup.
The parallelism seems like a good idea, but in practice in didn't improve the performances.

**Comment by trantienduchn** [comment]
> It's sad, but the parallel fetching execution of the mailboxes gives us nothing except an higher errors rate....

Accords to the image you posted. I don't see any error in gatling result, though.

**Comment by remk** [comment]
> > It's sad, but the parallel fetching execution of the mailboxes gives us nothing except an higher errors rate....
> 
> Accords to the image you posted. I don't see any error in gatling result, though.
>  Yes, but the last image is for the PR stripped from the parallelism optimization
> > The parallelism seems like a good idea, but in practice in didn't improve the performances.
> 
> That's sad

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: it's not an issue to accept an optional in a builder imho

[the remark this answers is not in the document]
Comment (remk) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: Ok then :+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 13]
```diff
@@ -121,31 +125,46 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
             .orElseGet(Throwing.supplier(() -> retrieveAllMailboxes(mailboxSession)).sneakyThrow());
     }
 
+    private MailboxFactory.MailboxBuilder getBuilder(
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: please don't, you defeat the goal of the builder: having readable argument passing at calling site

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java [hunk 14]
```diff
@@ -106,8 +112,10 @@ public MailboxBuilder builder() {
         return new MailboxBuilder(this);
     }
 
-    private Mailbox fromMessageManager(MessageManager messageManager, Optional<List<MailboxMetaData>> userMailboxesMetadata,
-                                                 MailboxSession mailboxSession) throws MailboxException {
+    private Mailbox fromMessageManager(MessageManager messageManager,
+                                       Optional<List<MailboxMetaData>> userMailboxesMetadata,
+                                       Optional<Quotas> preloadedDefaultUserQuotas,
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: I don't agree to make MailboxFactory more complex to handle a special case of a field being set with some preloaded values.
We could for example extract quota loading logic in a component and we take it there, opening the door to various implementations of that loader.
Or we can find any better design we want, but adding `if`s in MailboxFactory is not an option IMO.

**Comment by aduprat** [comment]
:green_apple: before CI restart

**Comment by remk** [comment]
test this please

**Comment by rouazana** [comment]
merged

**Pull request closed without merging**